mbuild/lattice.py

```diff
@@ -0,0 +1,401 @@
+            basis_dict = deepcopy(basis_vectors)
+
+        self.dimension = dimension
+        self.lattice_vectors = lattice_vectors
+        self.lattice_spacings = lattice_spacings
+        self.basis_vectors = basis_dict
+
+    def populate(self, x=None, y=None, z=None, compound_dict=None):
+        """
+        Expand lattice and create compound from lattice.
+
+        populate will expand lattice based on user input. The user must also
+        pass in a dictionary that contains the keys that exist in the
+        basis_dict. The corresponding Compound will be the full lattice
+        returned to the user.
+
+        If no dictionary is passed to the user, Dummy Compounds will be used.
+
+        Parameters
+        ----------
+        x : int, optional, default=None
+            How many iterations in the x direction.
+        y : int, optional, default=None
+            How many iterations in the y direction.
+        z : int, optional, default=None
+            How many iterations in the z direction.
+        compound_dict : dictionary, optional, default=None
+            Link between basis_dict and Compounds.
+
+        Exceptions Raised
+        -----------------
+        ValueError : incorrect x,y, or z values.
+        TypeError : incorrect type for basis vector
+
+        Call Restrictions
+        -----------------
+        Called after constructor by user.
+        """
+        if self.dimension == 3:
+            a = self.lattice_spacings[0]
+            b = self.lattice_spacings[1]
+            c = self.lattice_spacings[2]
+            if x is None:
+                x = 1
+            if y is None:
+                y = 1
+            if z is None:
+                z = 1
+            if x < 1 or y < 1 or z < 1:
+                raise ValueError('Incorrect populate value: X, Y, or Z is < 1.'
+                                 ' Cannot replicate unit cell less than 1')
+        elif self.dimension == 2:
+            a = self.lattice_spacings[0]
+            b = self.lattice_spacings[1]
+            if x is None:
+                x = 1
+            if y is None:
+                y = 1
+            if z is None:
+                pass
+            else:
+                raise ValueError('Z is defined although dimension is 2D')
+            if x < 1 or y < 1:
+                raise ValueError('Incorrect populate value: X or Y is < 1. '
+                                 ' Cannot replicate unit cell less than 1')
+        elif self.dimension == 1:
+            a = self.lattice_spacings[0]
+            if x is None:
+                x = 1
+            if y is None:
+                pass
+            else:
+                raise ValueError('Y is defined although dimension is 1D')
+            if z is None:
+                pass
+            if z is not None:
+                raise ValueError('Z is defined although dimension is 2D')
+            if x < 1:
+                raise ValueError('Incorrect populate value: X < 1. '
```

Comment:
I think there's probably some room to remove duplication in these 3 code paths. Do you need to unpack a, b, c? Can you do that later by reducing code re-use there?